job-runner/chain/chain.go

```diff
@@ -97,103 +135,197 @@ func (c *chain) NextJobs(jobId string) proto.Jobs {
+func (c *chain) IsDoneRunning() (done bool, complete bool) {
 	c.jcMux.RLock()
 	defer c.jcMux.RUnlock()
 
 	done = true
 	complete = true
 	pendingJobs := proto.Jobs{}
 
+	// If any jobs are still running, the chain isn't done or complete.
+	for _, job := range c.jobChain.Jobs {
+		if job.State == proto.STATE_RUNNING {
+			return false, false
+		}
+	}
+
 	// Loop through every job in the chain and act on its state. Keep
 	// track of the jobs that aren't running or in a finished state so
 	// that we can later check to see if they are capable of running.
-LOOP:
 	for _, job := range c.jobChain.Jobs {
 		switch job.State {
 		case proto.STATE_RUNNING:
-			// If any jobs are running, the chain can't be done
-			// or complete, so return false for both now.
+			// We should never reach this case because we previously check for running jobs.
 			return false, false
 		case proto.STATE_COMPLETE:
 			// Move on to the next job.
-			continue LOOP
-		case proto.STATE_FAIL:
-			// This failed job is part of a sequence that can be retried
-			if c.CanRetrySequence(job) {
+			continue
+		case proto.STATE_PENDING, proto.STATE_STOPPED:
+			// Treat stopped jobs as pending jobs because they may be retried
+			// (as in a resumed job chain)
+			pendingJobs = append(pendingJobs, job)
+		default:
+			// Any job that matches none of the above cases is failed
+			if c.canRetrySequence(job.Id) {
+				// This failed job is part of a sequence that can be retried.
 				return false, false
 			}
-		default:
-			// Any job that's not running, complete, or failed.
-			pendingJobs = append(pendingJobs, job)
 		}
 
-		// We can only arrive here if a job is not complete. If there
-		// is at least one job that is not complete, the whole chain is
-		// not complete. The chain could still be done, though, so we
-		// aren't ready to return yet.
+		// We can only arrive here if a job is not complete (pending or failed).
+		// If there is at least one job that is not complete, the whole chain
+		// is not complete. The chain could still be done, though, so we aren't
+		// ready to return yet.
 		complete = false
 	}
 
-	// For each pending job, check to see if all of its previous jobs
-	// completed. If they did, there's no reason the pending job can't run.
+	// For each pending job, check to see if it's ready to be run.
 	for _, job := range pendingJobs {
-		complete = false
-		allPrevComplete := true
-		for _, prevJob := range c.previousJobs(job.Id) {
-			if prevJob.State != proto.STATE_COMPLETE {
-				allPrevComplete = false
-				// We can break out of this loop if a single
-				// one of the previous jobs is not complete.
-				break
-			}
-		}
-
-		// If all of the previous jobs of a pending job are complete, the
-		// chain can't be complete because the pending job can still run.
-		if allPrevComplete == true {
-			return false, complete
+		if c.jobIsReady(job.Id) {
+			// If a job can be run, the chain can't be done.
+			done = false
+			break
```

Comment:
Might as well early return here if we're certain we know the final results. In such cases (early return), it prevents accidental code after the block, etc. Who knows what another dev might put after this block.

Reply:
Better yet, see comment above ^